WordPress/Classes/System/WordPressAppDelegate+openURL.swift

```diff
@@ -0,0 +1,162 @@
+import WordPressAuthenticator
+
+/// Custom pattern matcher used for our URL scheme
+private func ~=(pattern: String, value: URL) -> Bool {
+    return value.absoluteString.contains(pattern)
+}
+
+@objc extension WordPressAppDelegate {
+    internal func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool {
+        DDLogInfo("Application launched with URL: \(url)");
+        var returnValue = false
+
+        // 1. check if hockey can open this URL
+        var hockeyOptions: [String: Any] = [:]
+        for (key, value) in options {
+            hockeyOptions[key.rawValue] = value
+        }
+        if hockey.handleOpen(url, options: hockeyOptions) {
+            returnValue = true
+        }
+
+        // 2. check if this is a Google login URL
+        if WordPressAuthenticator.shared.handleGoogleAuthUrl(url,
+                                                             sourceApplication: options[UIApplication.OpenURLOptionsKey.sourceApplication] as? String,
+                                                             annotation: options[UIApplication.OpenURLOptionsKey.annotation]) {
+            returnValue = true
+        }
+
+        // 3. let's see if it's our wpcom scheme
+        guard url.absoluteString.hasPrefix(WPComScheme) else {
+            return returnValue
+        }
+
+        // this works with our custom ~=
+        switch url {
+        case "newpost":
+            returnValue = handleNewPost(url: url)
+        case "magic-login":
+            returnValue = handleMagicLogin(url: url)
+        case "viewpost":
+            returnValue = handleViewPost(url: url)
+        case "viewstats":
+            returnValue = handleViewStats(url: url)
+        case "debugging":
+            handleDebugging(url: url)
+        default:
+            break
+        }
+
+        return returnValue
+    }
+
+    private func handleMagicLogin(url: URL) -> Bool {
+        DDLogInfo("App launched with authentication link")
+        let allowWordPressComAuth = !AccountHelper.isDotcomAvailable()
+        guard let rvc = window.rootViewController else {
+            return false
+        }
+        return WordPressAuthenticator.openAuthenticationURL(url,
+                                                            allowWordPressComAuth: allowWordPressComAuth,
+                                                            fromRootViewController: rvc)
+    }
+
+    private func handleViewPost(url: URL) -> Bool {
+        guard let query = url.query,
+            let params = query.dictionaryFromQueryString(),
```

Comment:
We could use the system `URLComponents` / `URLQueryItems` types here instead, although I think they actually make the code worse...

```
private func handleViewPost(url: URL) -> Bool {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
            let queryItems = components.queryItems,
            queryItems.count > 0 else {
                return false
        }

        guard let blogIdValue = queryItems.first(where: { $0.name == "blogId" })?.value,
            let postIdValue = queryItems.first(where: { $0.name == "postId" })?.value,
            let blogId = Int(blogIdValue),
            let postId = Int(postIdValue) else {
                return false
        }

        WPTabBarController.sharedInstance()?.showReaderTab(forPost: NSNumber(value: postId),
                                                           onBlog: NSNumber(value: blogId))
        return true
    }
```

Reply:
It'd definitely be nice to drop the Obj-C of `dictionaryFromQueryString` 🤔

Reply:
We could probably write a small thing to wrap some of this up, as we need it in all the methods here.

Reply:
You could have a `URLComponents` extension to make it easier to read

```swift
extension URLComponents {
    var queryItemsDictionary: [String: String?] {
        guard let queryItems = queryItems else {
            return [:]
        }
        return queryItems.reduce([:], { (result, item) in
            var result = result
            result[item.name] = item.value
            return result
        })
    }
}
```

Reply:
I like this!

Reply:
I feel like there's still a nicer way to do that `reduce`. Maybe `Dictionary(uniqueKeysWithValues: queryItems.map({ ($0.name, $0.value) }))`? 🤷‍♂️